arwen/host/arwen.go

```diff
@@ -319,69 +316,55 @@ func (host *vmHost) RunSmartContractCreate(input *vmcommon.ContractCreateInput)
 	host.mutExecution.RLock()
 	defer host.mutExecution.RUnlock()
 
+	ctx, cancel := context.WithTimeout(context.Background(), executionTimeout)
+	defer cancel()
+
 	log.Trace("RunSmartContractCreate begin", "len(code)", len(input.ContractCode), "metadata", input.ContractCodeMetadata)
 
-	try := func() {
+	done := make(chan struct{})
+	go func() {
 		vmOutput = host.doRunSmartContractCreate(input)
-	}
-
-	catch := func(caught error) {
-		err = caught
-		log.Error("RunSmartContractCreate", "error", err)
-	}
+		close(done)
+	}()
 
-	TryCatch(try, catch, "arwen.RunSmartContractCreate")
-	if vmOutput != nil {
-		log.Trace("RunSmartContractCreate end", "returnCode", vmOutput.ReturnCode, "returnMessage", vmOutput.ReturnMessage)
+	select {
+	case <-done:
+		return
+	case <-ctx.Done():
+		err = arwen.ErrExecutionFailedWithTimeout
+		return
 	}
-
-	return
 }
 
 // RunSmartContractCall executes the call of an existing contract
 func (host *vmHost) RunSmartContractCall(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput, err error) {
 	host.mutExecution.RLock()
 	defer host.mutExecution.RUnlock()
 
-	log.Trace("RunSmartContractCall begin", "function", input.Function)
-
-	tryUpgrade := func() {
-		vmOutput = host.doRunSmartContractUpgrade(input)
-	}
+	ctx, cancel := context.WithTimeout(context.Background(), executionTimeout)
+	defer cancel()
 
-	tryCall := func() {
-		vmOutput = host.doRunSmartContractCall(input)
-	}
-
-	catch := func(caught error) {
-		err = caught
-		log.Error("RunSmartContractCall", "error", err)
-	}
-
-	isUpgrade := input.Function == arwen.UpgradeFunctionName
-	if isUpgrade {
-		TryCatch(tryUpgrade, catch, "arwen.RunSmartContractUpgrade")
-	} else {
-		TryCatch(tryCall, catch, "arwen.RunSmartContractCall")
-	}
-
-	return
-}
-
-// TryCatch simulates a try/catch block using golang's recover() functionality
-func TryCatch(try TryFunction, catch CatchFunction, catchFallbackMessage string) {
-	defer func() {
-		if r := recover(); r != nil {
-			err, ok := r.(error)
-			if !ok {
-				err = fmt.Errorf("%s, panic: %v", catchFallbackMessage, r)
-			}
+	log.Trace("RunSmartContractCall begin", "function", input.Function)
 
-			catch(err)
+	done := make(chan struct{})
+	go func() {
+		isUpgrade := input.Function == arwen.UpgradeFunctionName
```

Comment:
what happens, if, for some reason, this take too long, say, 10 seconds and the next call to the same contract, same function, same parameters comes to be executed?

`doRunSmartContractUpgrade`/`doRunSmartContractCall` will be ran in parallel. Valid also for `doRunSmartContractCreate`

Please add tests for the refactored functions

Reply:
there is no parallel execution with the same VM.

Reply:
if the `host.Runtime().FailExecution(err)` synchronously stop the execution, it is fine.

Reply:
Yes, `FailExecution` injects an imminent breakpoint in Wasmer. The contract will not be able to continue afterwards.